### ISSUE TITLE
lambdas cannot be defined in config.json?

### ISSUE THREAD

**Issue by OJFord** [xonsh/xonsh#1660]
[The docs](http://xon.sh/tutorial.html#customizing-the-prompt) state that:

> The $PROMPT variable can be a string, or it can be a function (of no arguments) that returns a string.

I believe this is only true when setting `$PROMPT = lambda: 'whatever'`, and not in the config file, where it would - I should have thought - be most useful to define it as a lambda anyway.

I'd be very glad to be mistaken, but I'm not sure what the syntax is or would be in order to define this in a JSON-parseable manner.

**Comment by scopatz** [comment]
Hi @OJFord, thanks for asking this.  If you want to use Python constructs, you need to put them in the run control `~/.xonshrc` file, not the config file.  The JSON config file is for truly static information. I hope that this helps

**Comment by OJFord** [comment]
Right, okay.

The `$PROMPT` I wanted define would just be simpler to define as a `lambda` - I wanted `(branch_name)` with the name in branch colour, but not the parentheses. The only way to achieve this in config.json seems to be:
```
{curr_branch: (}{branch_color}{curr_branch:{}}{NO_COLOR}{curr_branch:)}
```
which is fine, just not very readable.

An alternative solution would be making it possible to 'nest' brace-variables - my first attempt  looked like {curr_branch: ({branch_color}...`, which just prints the literal string as a prompt - but it's probably not worth the effort required.

I seem to recall an issue about the config.json disappearing completely in future version anyway? I'd be very much in favour of that, but I can't find the issue now... Is that still planned to happen?

**Comment by scopatz** [comment]
Nested lookup is a nice idea.  Please make a separate issue for that. 

> I seem to recall an issue about the config.json disappearing completely in future version anyway? I'd be very much in favour of that, but I can't find the issue now... Is that still planned to happen?

@adqm et al. do not like the JSON based configuration. I am not a huge fan of it myself (even though I implemented it). However, I have a hard time seeing it going away completely unless someone totally restructures how the wizard persists and depersists the information it stores.  The event system should be able to remove the need for a separate configuration step from a "this before that" loading perspective, but that hasn't yet been implemented in the core.

In short. JSON usage maybe minimized in the future, but it probably won't go away unless someone is willing to refactor some things in the wizard.

**Comment by scopatz** [comment]
We have changed the loading system significantly. Please feel free to reopen if needed
